packages/@lwc/template-compiler/src/transform/binding.ts

```diff
@@ -237,30 +255,35 @@ function buildAST(rootIRElement: IRElement | undefined): BindingASTNode | undefi
         children: [],
     });
 
-    const components = collectBindingASTComponents(rootIRElement);
+    const components = collectComponentsAndSlots(rootIRElement);
     components.forEach(component => {
         // Top-down path
-        const prunedPath = getPrunedPath(component, components);
+        const prunedPath = getPrunedPath(component);
         prunedPath.forEach((currentElement, index) => {
             // If the current element does not have a node representation in the AST.
             if (!astNodeMap.has(currentElement)) {
                 const parentElement = prunedPath[index - 1];
                 const parentASTNode = astNodeMap.get(parentElement)!;
-                if (currentElement.component) {
-                    const componentNode = transformToASTComponentNode(currentElement);
+                if (isComponent(currentElement) || isSlot(currentElement)) {
+                    let astNode;
+                    if (isComponent(currentElement)) {
+                        astNode = transformToComponentNode(currentElement);
+                    } else {
+                        astNode = transformToSlotNode(currentElement);
+                    }
                     if (hasDirective(currentElement)) {
                         // If the component has an inline directive, then create
                         // a "virtual" directive node and insert the directive
-                        // node between the parent and component node.
-                        const directiveNode = transformToASTNode(currentElement);
+                        // node between the parent and child node.
+                        const directiveNode = transformToDirectiveNode(currentElement);
                         parentASTNode.children.push(directiveNode);
```

Comment:
I see that we have tests for slots with directives. Would it be worth to add a test for a custom component with a directive as well?

Reply:
See `for-each-inline`